src/circuits/blake2s/blake2s_comp.tcc

```diff
@@ -0,0 +1,178 @@
+    const libsnark::digest_variable<FieldT> &output,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , input_block(input_block)
+    , output(output)
+{
+    // Allocate and format the 16 input block variable
+    for (size_t i = 0; i < 16; i++) {
+        block[i].allocate(
+            pb, 32, FMT(this->annotation_prefix, " block_%zu", i));
+    }
+
+    // Setup constants, parameters and first state
+    ZERO.allocate(pb, FMT(this->annotation_prefix, " ZERO"));
+    BLAKE2s_256_comp<FieldT>::setup_constants();
+    BLAKE2s_256_comp<FieldT>::setup_h();
+
+    // Allocate the states
+    for (size_t i = 0; i < rounds + 1; i++) {
+        for (size_t j = 0; j < 16; j++) {
+            v[i][j].allocate(
+                this->pb,
+                32,
+                FMT(this->annotation_prefix, " v_%zu", i * rounds + j));
+        }
+    }
+    for (size_t i = 0; i < rounds; i++) {
+        for (size_t j = 0; j < 16; j++) {
+            v_temp[i][j].allocate(
+                this->pb,
+                32,
+                FMT(this->annotation_prefix, " v_temp_%zu", i * rounds + j));
+        }
+    }
+
+    // Allocate output bytes (before swapping endianness and appending)
+    for (size_t i = 0; i < 8; i++) {
+        output_bytes[i].allocate(
+            pb, 32, FMT(this->annotation_prefix, " output_byte_%zu", i));
+    }
+
+    setup_gadgets();
+};
+
+template<typename FieldT>
+void BLAKE2s_256_comp<FieldT>::generate_r1cs_constraints(
+    const bool ensure_output_bitness)
+{
+    libff::UNUSED(ensure_output_bitness);
+
+    libsnark::generate_r1cs_equals_const_constraint<FieldT>(
+        this->pb, ZERO, FieldT::zero(), FMT(this->annotation_prefix, " ZERO"));
+
+    for (size_t i = 0; i < rounds; i++) {
+        for (auto &gadget : g_arrays[i]) {
+            gadget.generate_r1cs_constraints();
+        }
+    }
+
+    for (auto &gadget : xor_vector) {
+        gadget.generate_r1cs_constraints();
+    }
+};
+
+template<typename FieldT> void BLAKE2s_256_comp<FieldT>::generate_r1cs_witness()
+{
+    // Format two 256 bit long big endian inputs into one 512 long little endian
+    // input (with padding if necessary)
+    size_t input_size = input_block.bits.size();
+    // We do not use block_size because the value might not be entered
+    //(c.f. block_variable<FieldT>::block_variable(protoboard<FieldT> &pb,
+    //                                   const
+    //                                   std::vector<pb_variable_array<FieldT> >
+    //                                   &parts, const std::string
+    //                                   &annotation_prefix) )
+
+    // Push the block variable in local to be swapped
+    std::vector<FieldT> padded_input;
+    for (size_t i = 0; i < input_size; i++) {
+        padded_input.push_back(this->pb.val(input_block.bits[i]));
+    }
+
+    // [SANITY CHECK] Pad if necessary
+    for (size_t i = 0; i < 512 - input_size; i++) {
```

Comment:
512 is the block_size. Please replace by the constant name associated to it (ie: `BLAKE2s_block_size`)